src/stan/analyze/mcmc/compute_effective_sample_size.hpp

```diff
@@ -1,98 +1,184 @@
-      }
+namespace analyze {
 
-      Eigen::Matrix<Eigen::VectorXd, Eigen::Dynamic, 1> acov(num_chains);
-      Eigen::VectorXd chain_mean(num_chains);
-      Eigen::VectorXd chain_var(num_chains);
-      for (int chain = 0; chain < num_chains; ++chain) {
-        Eigen::Map<const Eigen::Matrix<double, Eigen::Dynamic, 1>>
-          draw(draws[chain], sizes[chain]);
-        math::autocovariance<double>(draw, acov(chain));
-        chain_mean(chain) = draw.mean();
-        chain_var(chain) = acov(chain)(0) * num_draws / (num_draws - 1);
-      }
+  /**
+   * Returns the effective sample size for the specified parameter
+   * across all kept samples.
+   *
+   * See more details in Stan reference manual section "Effective
+   * Sample Size". http://mc-stan.org/users/documentation
+   *
+   * Current implementation assumes chains are all of equal size and
+   * draws are stored in contiguous blocks of memory.
+   *
+   * @param std::vector stores pointers to arrays of chains
+   * @param std::vector stores sizes of chains
+   * @return effective sample size for the specified parameter
+   */
+  inline
+  double compute_effective_sample_size(std::vector<const double*> draws,
+                                       std::vector<size_t> sizes) {
+    int num_chains = sizes.size();
+
+    // need to generalize to each jagged draws per chain
+    size_t num_draws = sizes[0];
+    for (int chain = 1; chain < num_chains; ++chain) {
+      num_draws = std::min(num_draws, sizes[chain]);
+    }
+
+    Eigen::Matrix<Eigen::VectorXd, Eigen::Dynamic, 1> acov(num_chains);
+    Eigen::VectorXd chain_mean(num_chains);
+    Eigen::VectorXd chain_var(num_chains);
+    for (int chain = 0; chain < num_chains; ++chain) {
+      Eigen::Map<const Eigen::Matrix<double, Eigen::Dynamic, 1>>
+        draw(draws[chain], sizes[chain]);
+      autocovariance<double>(draw, acov(chain));
+      chain_mean(chain) = draw.mean();
+      chain_var(chain) = acov(chain)(0) * num_draws / (num_draws - 1);
+    }
 
-      double mean_var = chain_var.mean();
-      double var_plus = mean_var * (num_draws - 1) / num_draws;
-      if (num_chains > 1)
-        var_plus += math::variance(chain_mean);
-      Eigen::VectorXd rho_hat_s(num_draws);
-      rho_hat_s.setZero();
-      Eigen::VectorXd acov_s(num_chains);
+    double mean_var = chain_var.mean();
+    double var_plus = mean_var * (num_draws - 1) / num_draws;
+    if (num_chains > 1)
+      var_plus += math::variance(chain_mean);
+    Eigen::VectorXd rho_hat_s(num_draws);
+    rho_hat_s.setZero();
+    Eigen::VectorXd acov_s(num_chains);
+    for (int chain = 0; chain < num_chains; ++chain)
+      acov_s(chain) = acov(chain)(1);
+    double rho_hat_even = 1.0;
+    rho_hat_s(0) = rho_hat_even;
+    double rho_hat_odd = 1 - (mean_var - acov_s.mean()) / var_plus;
+    rho_hat_s(1) = rho_hat_odd;
+    // Geyer's initial positive sequence
+    size_t s = 1;
+    for (; s < (num_draws - 4) && (rho_hat_even + rho_hat_odd) > 0;
```

Comment:
Personally I think that hacking a for loop instead of using a while loop here makes the logic a bit harder to follow but not so much that I'd push back if you prefer the for loop.